File: All Lab Tasks/Lab5_Bundlers/src/Bundle.java
import java.util.ArrayList;
import java.util.List;

public abstract class Bundle implements IProduct {
    protected String name;
    protected String description;
    protected double discount;

    List<IProduct> products = new ArrayList<>();

    public Bundle(String name, String description, double discount) {
        this.name = name;
        this.description = description;
        this.discount = discount;
    }


    public abstract void createBundle();

    public List<IProduct> getProducts() {
        return this.products;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public double getPrice() {
        double total_price = 0;
        for (IProduct product : getProducts()) {
            total_price += product.getPrice();
        }
        return total_price - discount;
    }

    @Override
    public String getDescription() {
        return description;
    }

    @Override
    public void displayProductInfo() {
        System.out.println("Bundle: " + name + " | Discount: $" + discount + " | Description: " + description);
        for (IProduct product : getProducts()) {
            product.displayProductInfo();
        }
        System.out.println("Total Bundle Price (After Discount): $" + getPrice());
    }
}

File: Lab5_Bundlers/Lab5_Bundlers/Bundle.cs
using System;
using System.Collections.Generic;

namespace Lab5_Bundlers
{
    // Composite - Bundle
    public abstract class Bundle : IProduct
    {
        private string _name;
        private decimal _discount; // Discount applied to the total bundle price
        protected List<IProduct> _products = new List<IProduct>();

        public Bundle(string name, decimal discount)
        {
            _name = name;
            _discount = discount;
        }

        public void AddProduct(IProduct product)
        {
            _products.Add(product);
        }

        public void RemoveProduct(IProduct product)
        {
            _products.Remove(product);
        }

        public string GetName()
        {
            return _name;
        }

        public decimal GetPrice()
        {
            decimal total = 0;
            foreach (var product in _products)
            {
                total += product.GetPrice();
            }
            // Apply discount to the total price of the bundle
            return total * (1 - _discount);
        }

        public void DisplayProductDetails()
        {
            Console.WriteLine($"Bundle: {_name}, Total Price after Discount: BDT {GetPrice():0.00}");
            Console.WriteLine("Contains:");
            foreach (var product in _products)
            {
                product.DisplayProductDetails();
            }
        }

        protected abstract List<IProduct> BuildBundle();
    }
}
